onnx/checker.cc

```diff
@@ -389,31 +401,33 @@ void check_graph(
     check_value_info(value_info, ctx);
   }
 
-  std::unordered_set<std::string> output_names{};
-  // Inherit values avaiailable in outer scope
+  // Inherit values available in outer scope
   // Note that we do not allow shadowing, so the presence of an already-defined
   // name is always an error.
+  LexicalScopeContext lex_ctx{parent_lex};
+
   for (const auto& value_info : graph.input()) {
-    if (output_names.count(value_info.name())) {
+    // TODO: If shadowing isn't allowed, this should maybe use
```

Comment:
check line 405, we don't allow shadowing

Reply:
Doesn't that happens after the graph (or subgraph) inputs are checked? So in the case of a subgraph, a value in the parent graph could be shadowed by a subgraph input.

Reply:
@houseroad : taking a step back: the existing version of the checker allows shadowing in the case of graph-inputs (as Scott mentions). However, it does look slightly odd, considering that shadowing is not allowed for node outputs inside the subgraph. I believe Scott's current version preserves the existing behavior, but if it seems appropriate we could change the behavior to forbid shadowing for the subgraph inputs also. (The concern is whether it will break some existing model; @skottmckay : I assume that the related bug in ORT implies that no model currently using ORT could have used this feature successfully?)

Reply:
Regarding the bug in ORT, the model could have run but it would not have been correct as it would have been using the incorrect value (outer scope value instead of the subgraph input value).